File: LeetCode/May challenge/Day_7.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        def f(root, v, p):
            if root.val == v:
                _d = 1
                _p = p
                return _d, _p
            
            if root.left!=None:
                l_d, l_p = f(root.left, v, root.val)
                if l_d!=None:
                    return l_d+1, l_p
                
            if root.right!=None:
                r_d, r_p = f(root.right, v, root.val)
                if r_d!=None:
                    return r_d+1, r_p             
            return None, p
			
        _x = f(root, x, None)
        _y = f(root, y, None)
        return (_x[0] == _y[0]) and (_x[1] != _y[1])

File: LeetCode/May challenge/Day 7/Day_7.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int parent(TreeNode* node,int n)
    {
        if(node==NULL) 
            return -1;
        else 
            if((node->left!=NULL && node->left->val==n) || (node->right!=NULL && node->right->val==n)) 
                return node->val;
             else
                return max(parent(node->left,n),parent(node->right,n));
    }
    int level(TreeNode* node, int n, int d)
    {
        if(node==NULL) 
            return 0;
        else if(node->val==n)
            return d;
        else 
            return max(level(node->left,n,d+1),level(node->right,n,d+1));
    }
    
    bool isCousins(TreeNode* root, int x, int y) {
        int levelx= level(root,x,0);
        int levely= level(root,y,0);
        if(levelx!=levely)
            return false;
        
        if(parent(root,x)==parent(root,y))
            return false;
        else
            return true;
    }
};
